File: 이것이 취업을 위한 코딩 테스트다/그래프_이론/팀_결성/팀_결성.cpp
#include <iostream>

using namespace std;

int find_team(int* team, int x) {
  if (team[x] != x) {
    team[x] = find_team(team, team[x]);
  }

  return team[x];
}

void union_team(int* team, int a, int b) {
  a = find_team(team, a);
  b = find_team(team, b);

  if (a < b) {
    team[b] = a;
  } else {
    team[a] = b;
  }
}

int main() {
  int team[100001];
  int n, m;
  cin >> n >> m;

  for (int i = 0; i <= n; i++) {
    team[i] = i;
  }

  for (int i = 0; i < m; i++) {
    int type, a, b;
    cin >> type >> a >> b;

    if (type) {
      if (find_team(team, a) == find_team(team, b)) {
        cout << "YES\n";
      } else {
        cout << "NO\n";
      }
    } else {
      union_team(team, a, b);
    }
  }

  return 0;
}

File: 이것이 취업을 위한 코딩 테스트다/그래프_이론/팀_결성/팀_결성.py
def find_team(team, x):
  if team[x] != x:
    team[x] = find_team(team, team[x])

  return team[x];

def union_team(team, a, b):
  a = find_team(team, a)
  b = find_team(team, b)

  if a < b:
    team[b] = a
  else:
    team[a] = b

n, m = map(int, input().split())
team = [0] * (n + 1)

for i in range(n + 1):
  team[i] = i

for _ in range(m):
  type, a, b = map(int, input().split())

  if type == 1:
    if find_team(team, a) == find_team(team, b):
      print("YES")
    else:
      print("NO")
  else:
    union_team(team, a, b)